### ISSUE TITLE
Align with the proposed WebAssembly/ESM integration specification

### ISSUE THREAD

**Issue by littledan** [webpack/webpack#8144]
_Update: On further investigation, it seems that there's a very significant mismatch between webpack Wasm semantics and the early draft proposed standard, in terms of when Wasm instantiation happens. I'm happy to have the help of the webpack team here, with your experience this space and success building a working system. Let's keep discussing this to figure out what semantics should be standardized._

## Feature request

The WebAssembly CG is working on standardizing the integration of WebAssembly modules and ES Modules. The [proposal's semantics](https://github.com/WebAssembly/esm-integration/blob/e098188221887871410616ba1e186740c9d4aa86/proposals/esm-integration/README.md) are very similar to what webpack's experimental WebAssembly support. However, there may be some minor differences. This issue is intended to be an umbrella bug to track the general effort to add tests against the new proposal and semantic changes to match it.

**What is the expected behavior?**

That webpack's WebAssembly module support would match the new specification.

**What is motivation or use case for adding/changing the behavior?**

- All documentation about wasm and ESM (the spec itself, MDN, other developer resources) could match up with webpack's implementation
- Bring compatibility among bundlers, to make things easier for programmers
- Make native modules and webpack behave the same, in case the same code may be used both with webpack and as a native module

**How should this be implemented in your opinion?**

Let's start with tests. Long-term, it'd be best if tests are part of web-platform-tests and imported into webpack; that's a bit of an unsolved problem for now, though, so the initial plan is to develop tests directly against webpack in a PR.

Test plan:
- Convert the remaining .wasm tests to .wat for easier development
- Exports of globals from Wasm are imported as WebAssembly.Global
objects, not Numbers (this looks likely to be an existing spec violation)
- Imports of Numbers become only immutable, not mutable, Global objects,
and undergo rounding based on the declared type
- A WebAssembly.Table, Memory or Global instance, or exported Wasm
function, exported JS -> Wasm -> JS, comes out with the same identity
- A cycle Wasm <-> JS, with Wasm on top, leads to a TDZ error for
Memory, Global or Table, but hoisted functions work, and Wasm things are
all initialized by the time the JS top-level statement runs
- A cycle JS <-> Wasm, with JS on top, lets Wasm's start function access
Memory, Global, Table and functions, but using Wasm exports too early
leads to TDZ
- Variables imported JS -> Wasm are snapshotted (whether you overwrite a
bare Number or function, or an entire Memory, Global or Table object
which isn't replaced when the variable containing it is overwritten)
- Memory, Table and mutable Global can be mutated by JS code that's
called out from Wasm code, and the mutations are observed from Wasm
- LinkError results if there is a type mismatch when importing into
WebAssembly (whether it's from JS or Wasm)
- Circular Wasm modules result in an error before any start function
executes
- All Wasm modules in the module graph are validated before any start
function runs or JS modules are evaluated; failing validation throws a
CompileError
- The same Wasm module can be imported multiple times (both from JS and
Wasm) and only one copy of it is made (1 start function evaluation, 1
memory/table/global identity, etc)
- The name (number) of exported functions, and the enumeration order of
exports overall, matches the JS API specification
- Wasm modules can be imported from an ordinary import inside an inline
a JS module and a dynamic import from either a script or a JS module
- Some parts of the semantics will probably be different in webpack and with native ESM; these tests might only be landed in wpt and not in webpack:
    - The return value of import() for a Wasm module is a
module namespace exotic object, similar to that of native modules
(WebPack will need to take shortcuts here)
    - The same resource can be used both as a Wasm module as
well as through instantiateStreaming, and there's no observable
caching/sharing between these
    - The choice between interpreting as JS and Wasm is
based on the mimetype and nothing else; many URL types work (blob, data,
served from SW), CSP is checked appropriately, the modern CORS settings
are used for the fetch, etc.

**Are you willing to work on this yourself?**

I plan to start implementing the test plan described above. (I wouldn't mind help here, cc @ms2ger @xtuc) I am not sure when I would have time to implement the changes in webpack; definitely not before the end of this year.

**Comment by xtuc** [comment]
Here are a few points I have in mind:

> Convert the remaining .wasm tests to .wat for easier development

Some are intentially binary to test against various LEB128 encoding, but I have some on my side too, I think we could just remove/move them from here.

> A WebAssembly.Table, Memory or Global instance, or exported Wasm function, exported JS -> Wasm -> JS, comes out with the same identity

What does identity mean here? The same function object instance?

We already discussed about that because it's an optimization, we could directly wire up the function in JS and avoid calling in wasm entirely.

> Variables imported JS -> Wasm are snapshotted (whether you overwrite a bare Number or function, or an entire Memory, Global or Table object which isn't replaced when the variable containing it is overwritten)

Isn't that already handle by WebAssembly's instantiation which does a copy of the values? Versus being a live-binding.

> Circular Wasm modules result in an error before any start function executes

Which means an error at compilation because the start function is executed in the same tick than the instantiation.

> The same Wasm module can be imported multiple times (both from JS and Wasm) and only one copy of it is made (1 start function evaluation, 1 memory/table/global identity, etc)

Modules are identified by their URL, right?

> I wouldn't mind help here, cc @Ms2ger @xtuc

As I said, I will be able to work on this, and before the end of the year.

**Comment by sokra** [comment]
> * Imports of Numbers become only immutable, not mutable, Global objects,
>   and undergo rounding based on the declared type

Does this mean?

| wasm type \ js runtime type -> | Number | WebAssembly.Global |
|---|---|---
| global i32 | snapshot | snapshot |
| global mut i32 | Error (at runtime) | binding |

> * A WebAssembly.Table, Memory or Global instance, or exported Wasm
>   function, exported JS -> Wasm -> JS, comes out with the same identity
> * A cycle Wasm <-> JS, with Wasm on top, leads to a TDZ error for
>   Memory, Global or Table, but hoisted functions work, and Wasm things are
>   all initialized by the time the JS top-level statement runs

We will fail these, because it's not possible to import table or memory from JS into WASM

> * Memory, Table and mutable Global can be mutated by JS code that's
>   called out from Wasm code, and the mutations are observed from Wasm

Do the current WASM implementation already support importing mutable globals? This was not part of the MVP, was it?

> * Circular Wasm modules result in an error before any start function
>   executes

This seems like an unnecessary restriction to me and is not the case of ESM. We will fail this because this is allowed.

> * The return value of import() for a Wasm module is a
>   module namespace exotic object, similar to that of native modules
>   (WebPack will need to take shortcuts here)

In webpack it's not so exotic. It's an object with `Symbol.toStringTag = "Module"`... The same for ESM.

**Comment by sokra** [comment]
> * Exports of globals from Wasm are imported as WebAssembly.Global
>   objects, not Numbers (this looks likely to be an existing spec violation)

We pass along the value of the property in `WebAssembly.Instance.exports`.

Seem like this was a recent change in the spec. The JS API now exports `WebAssembly.Global` for globals. I actually wonder why this change was made because it's a breaking change for non-mutable globals. We may need to update our test suite, when this update 
has been made to Node.js. https://github.com/WebAssembly/spec/commit/d8f775eed3645e689580610d101106cd1404b637#diff-ef3ad1622710e5adc119fbe2af30971cR376

> * Imports of Numbers become only immutable, not mutable, Global objects,
>   and undergo rounding based on the declared type

Sounds like importing a number as mutable global is allowed, but mutation the global will lead to an runtime error inside the wasm.

**Comment by sokra** [comment]
![image](https://user-images.githubusercontent.com/1365881/46597370-23fbf500-cae0-11e8-9684-9d3726270554.png)

Here we also divert from the spec. We use live-bindings for functions instead of a snapshot.

We should be able to change this by snapshotting all imported functions on initialization.

**Comment by sokra** [comment]
![image](https://user-images.githubusercontent.com/1365881/46597680-77227780-cae1-11e8-8a1c-ebede7f712c4.png)

Here we also divert. We apply the same rules as for `wasm imports <- JS exports`. Seems like the spec expects a different behavior here for reexports. We need a special-case here and resolve reexports. Seem like this is only possible with ESM as it statically declares reexports. When the same module is transpiled to CJS it would break.

**Comment by sokra** [comment]
There is no special case for `JS imports <- wasm re-exports <- JS exports` so I assume functions are snapshotted here at the time of wasm evaluation.

cc @linclark

**Comment by sokra** [comment]
![image](https://user-images.githubusercontent.com/1365881/46597992-aeddef00-cae2-11e8-84d7-c835c3259059.png)

The spec is not clear here:

* Did the init expression for globals already run at the time of JS evaluation or does it run when the start function runs?
* Same question for the `data` and `elem` segment?
* Is reading globals, memory and tables allowed in JS evaluation or will it lead to a TDZ Error?

The current webpack implementation does it this way:

* `data` and `elem` run at instantiation, which happens before the JS evaluation
* not-imported globals and globals not referencing imported globals (get_global in init) also run at instantiation
* imported globals and globals referencing imported globals run at evaluation

This is a bit inconsistent and can not change since it's a kind of workaround for importing values from JS into WASM. Once this is no longer allowed all global init can run at instantiation (and we connect wasm-to-wasm global imports directly).

Note: When init, data and elem should run at evaluation, they would be uninitialized in JS evaluation. It would be very in-efficient for webpack to delay the data section until the evaluation phase. A TDZ error would be possible. Using a TDZ Error would make it irrelevant when these sections run since it's no longer visible from outside.

**Comment by sokra** [comment]
![image](https://user-images.githubusercontent.com/1365881/46600193-b5239980-cae9-11e8-984b-466eb89bb752.png)

The spec often claims that functions are already initialized before evaluation. This is not necessary true for all functions exported by JS. It's only true for FunctionDeclarations. It's not true for i. e. `export const fn = () => {}`. Since the spec snapshots exports in the Instantiation phase these exports would result in errors.

It might be better to snapshot functions in the Evaluation phase where these exports are available. But this is more difficult to implement with the current WASM JS API (trampoline functions).

Other option is to use live bindings for functions, which also results in trampoline functions.

The spec also claims that all other exports have the value `undefined` which is also not true. They can also be in TDZ state. i. e. `export const a = 1`. Reading it will result in a TDZ Error. This results in the same behavior, but should be mentioned.

**Comment by sokra** [comment]
![image](https://user-images.githubusercontent.com/1365881/46601770-7f34e400-caee-11e8-9a29-3834775d61bb.png)

This seem to be needed because functions are snapshotted. When functions are threaded as live bindings this would be possible.

**Comment by littledan** [comment]
@xtuc 

> Some are intentially binary to test against various LEB128 encoding, but I have some on my side too, I think we could just remove/move them from here.

Seems like my test plan was poorly worded. If that's what you were testing for, that's great, no particular reason to remove it. I'd just like to get full test coverage of the things discussed above in text 

>> A WebAssembly.Table, Memory or Global instance, or exported Wasm function, exported JS -> Wasm -> JS, comes out with the same identity
>
> What does identity mean here? The same function object instance?
>
> We already discussed about that because it's an optimization, we could directly wire up the function in JS and avoid calling in wasm entirely.

To be clear, the "same identity" I'm talking about for functions just applies if we're talking about an *exported Wasm function*, which is then re-imported and re-exported. If you import a JS function into WebAssembly and then export it again, its behavior changes in that arguments are coerced to Wasm types according to how the function is imported. According to the Wasm-JS API specification, this isn't quite an optimization but rather exactly how the behavior is defined.

> Isn't that already handle by WebAssembly's instantiation which does a copy of the values? Versus being a live-binding.

Yes, it is (as is the above identity). The idea is just to write a test that verifies that everything went well.

>> Circular Wasm modules result in an error before any start function executes
>
> Which means an error at compilation because the start function is executed in the same tick than the instantiation.

An error at compilation sounds like a good solution to me. Is there a test for this yet?

> Modules are identified by their URL, right?

Yes, AFAIK (that's HTML's responsibility). I don't know the details of how paths different URLs/redirects might be identified with each other; presumably this is the kind of thing which might reasonably differ between webpack and the Web (given you already add extensions, which it doesn't look like the web will do).

@sokra 

> Does this mean?

Yes, in particular: If you pass in a Number, it will become a const Global, per step 4.5 of [the JS instantiate algorithm](http://webassembly.github.io/spec/js-api/index.html#instantiate-a-webassembly-module). Then, in [the WebAssembly instantiate algorithm](https://webassembly.github.io/spec/core/exec/modules.html#exec-instantiation), step 4, it's checked whether the types match, which will not be the case if a mutable global is expected.

> We will fail these, because it's not possible to import table or memory from JS into WASM

I'm wondering, is there anything blocking webpack from supporting importing table or memory from JS into Wasm?

> Do the current WASM implementation already support importing mutable globals? This was not part of the MVP, was it?

It wasn't part of the initial MVP, but it's part of the 1.0 specification (rushed to minimize the compat risk), and shipping in Firefox and Chrome.

>> This seems like an unnecessary restriction to me and is not the case of ESM. We will fail this because this is allowed.
>
> This seems like an unnecessary restriction to me and is not the case of ESM. We will fail this because this is allowed.

Indeed, it looks like you put in a ton of work to make circularities between multiple WebAssembly modules work. @rossberg, @lukewager and @linclark can fill in more details, but my understanding is that one big reason to not support circularities is in relation to the [gc proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md), which adds a way for modules to import types from other modules, and which would not map very well to circularities. (However, in JS, circular modules with classes extending each other don't really work out well either, as @concavelenz and others have pointed out... I don't quite understand why types are worse.) I have a feeling we'll need to spend some more time on this circularity, at the very least to document the motivation for this prohibition better, but possibly also to consider alternatives.

> In webpack it's not so exotic. It's an object with Symbol.toStringTag = "Module"... The same for ESM.

That sounds like a good solution to me.

((I'll follow up soon with responses to the later comments on the thread.))

**Comment by rossberg** [comment]
Re cyclic linking: there are at least four features in JavaScript that its cyclical linking ability depends on: (1) runtime checks, (2) TDZ failures/undefined values, (3) non-local "hoisting" of functions, (4) the complete absence of type definitions. These are costs that JS has already sunk, but for other languages, including Wasm, they are not necessarily feasible or acceptable (and even JS doesn't always swallow them, consider mutually recursive classes).

For that reason, you shouldn't assume that linking between JS and other languages can be as liberal as linking between JS and JS. I believe Allan WB has pointed out something similar in an earlier thread on the subject over on the TC39 github.

**Comment by littledan** [comment]
Probably it's not time yet to start landing changes to webpack based on the proposed specification. It looks like a number of @sorka's comments above are based on the version at master, while I'm betting on [this PR](https://github.com/WebAssembly/esm-integration/pull/13) to land.

Some more thoughts on cyclic Wasm modules at https://github.com/WebAssembly/esm-integration/issues/17 .

**Comment by sokra** [comment]
I'm betting against [that PR](https://github.com/WebAssembly/esm-integration/pull/13). It makes a fundamental change that make it in my opinion incompatible with the JS API. Snapshoting can't happen in the evaluation phase, because when using the JS API the `importsObject` need to be passed on `instantiate` and this is an async function. Evaluation phase is running sync in a single microtask. One could change the ESM spec, but I don't think it's worth a breaking change here.

We actually implemented it this way in the first prototype using sync instantiate while evaluation, but using sync instantiate was not acceptable. There is also an issue about that by Lin.

**Comment by littledan** [comment]
> Snapshoting can't happen in the evaluation phase, because when using the JS API the importsObject need to be passed on instantiate and this is an async function. Evaluation phase is running sync in a single microtask. 

This is an interesting point that I hadn't considered. But I don't understand--why do you expect evaluation to happen all in a single microtask?

**Comment by xtuc** [comment]
@littledan that's how it works currently, the instantiation until the start func gets evaluated is done in a single microtask.

**Comment by littledan** [comment]
OK, we'll have to think more about this issue about whether it would be OK to do Wasm instantiation asynchronously during the ESM evaluation phase. To respond to your points:

1. You're right. If we go down this path, we'll have to make further JS spec changes to allow asynchronous-ness in the Evaluation step. These changes will be needed for [top-level await](https://github.com/tc39/proposal-top-level-await) as well, so we should figure out how to do them in a common way. cc @MylesBorins @GeorgNeis

2. We might even allow other tasks, not just microtasks, in the middle, if compilation happens during evaluation and it takes significant time. What's the problem, though? What sorts of application logic would be broken by this? (Presumably this brokenness would also affect top-level await similarly.)

3. I'm not sure the performance impact would be so bad:

> When wasm instantiation happens in the instantiation phase, implementations are free to do some instantiations in parallel, since this is not visible at runtime. Performance benefit.

Some speculation from someone who has never worked on a Wasm implementation:

Even if Wasm instantiation happens in the evaluation phase, implementations may still be able to do some parts of starting up the Wasm module, especially compilation, in parallel in practice. Some parallelism could be based on all dependencies being satisfied (in which case, everything can proceed completely correctly until the start function, which would have to be run at the appropriate linear time); other parallelism could be speculative and possibly invalidated by later modules (e.g., if memory is imported and the module specialization depends on the memory type; the speculative compilation could even be kicked off in the instantiation phase).

I'll discuss this with implementers next week at TPAC to see what they think.

**Comment by sokra** [comment]
> 2\. What sorts of application logic would be broken by this?

I could imaging a module connecting to a (web)socket and emitting events for each message incoming. When used this way:

``` js
import socket from "./socket";
import method from "other-unrelated-module";

socket.on("message", msg => console.log(msg));
```

Code is working fine so far. Now when `other-unrelated-module` introduce some top-level-await or wasm which is async at evaluation time, you may loose some messages, because there is something happening in-between.

> 2\. Presumably this brokenness would also affect top-level await similarly.

yes, I'm also against that.

> Even if Wasm instantiation happens in the evaluation phase, implementations may still be able to do some parts of starting up the Wasm module, especially compilation, in parallel in practice. Some parallelism could be based on all dependencies being satisfied (in which case, everything can proceed completely correctly until the start function, which would have to be run at the appropriate linear time); other parallelism could be speculative and possibly invalidated by later modules (e.g., if memory is imported and the module specialization depends on the memory type; the speculative compilation could even be kicked off in the instantiation phase).

This is something browser might be able to do, but not something we (as webpack) can do with the current WASM JS API.

**Comment by littledan** [comment]
For what it's worth, I posted an early draft of a specification for the semantics we've been talking about at https://littledan.github.io/esm-integration/#esm-integration .

> This is something browser might be able to do, but not something we (as webpack) can do with the current WASM JS API.

That's right, it's not possible directly, but I wonder if there might be some kind of transform strategy to achieve similar semantics. (Sorry, I don't have a particular suggestion.)

**Comment by littledan** [comment]
I'd like to revive this topic. I've been working on the documentation and specification text for WebAssembly/ESM integration, and I plan to propose it for Phase 2 at an upcoming WebAssembly CG meeting. I'd be interested in more feedback from webpack here. I've expanded on the motivation for these semantics, and added in the detail about the relationship with top-level await (which you can also see reflected in the proposed spec text). See https://github.com/WebAssembly/esm-integration/blob/master/proposals/esm-integration/README.md for more details.

**Comment by evilebottnawi** [comment]
bump

**Comment by littledan** [comment]
I want to note: Unfortunately, I'm really sorry to report that it's not certain whether browsers will implement what webpack has done here. Specifically, browsers *might* require import sites to mark WebAssembly modules as such. See https://github.com/littledan/proposal-module-attributes/issues/19 for further discussion--your thoughts and comments there may be welcome.

**Comment by xtuc** [comment]
I propose that we close this issue until we have a concret plan.
